aptos-move/framework/aptos-framework/sources/nonce_validation.move

```diff
@@ -0,0 +1,256 @@
+            );
+        };
+        let bucket = table::borrow_mut(&mut nonce_history.nonce_table, bucket_index);
+
+        let existing_exp_time = bucket.nonce_to_exp_time_map.get(&nonce_key);
+        if (existing_exp_time.is_some()) {
+            let existing_exp_time = existing_exp_time.extract();
+
+            // If the existing (address, nonce) pair has not expired, return false.
+            if (existing_exp_time >= current_time) {
+                return false;
+            };
+
+            // We maintain an invariant that two transaction with the same (address, nonce) pair cannot be stored
+            // in the nonce history if their transaction expiration times are less than `GARBAGE_COLLECTION_INTERVAL_SECS_AFTER_TXN_EXPIRATION`
+            // seconds apart.
+            if (txn_expiration_time <= existing_exp_time + GARBAGE_COLLECTION_INTERVAL_SECS_AFTER_TXN_EXPIRATION) {
+                return false;
+            };
+
+            // If the existing (address, nonce) pair has expired, garbage collect it.
+            bucket.nonce_to_exp_time_map.remove(&nonce_key);
+            bucket.nonces_ordered_by_exp_time.remove(&NonceKeyWithExpTime {
+                txn_expiration_time: existing_exp_time,
+                sender_address,
+                nonce,
+            });
+        };
+
+        // Garbage collect upto MAX_ENTRIES_GARBAGE_COLLECTED_PER_CALL expired nonces in the bucket.
+        let i = 0;
+        while (i < MAX_ENTRIES_GARBAGE_COLLECTED_PER_CALL && !bucket.nonces_ordered_by_exp_time.is_empty()) {
+            let (front_k, _) = bucket.nonces_ordered_by_exp_time.borrow_front();
+            // We garbage collect a nonce after it has expired and the GARBAGE_COLLECTION_INTERVAL_SECS_AFTER_TXN_EXPIRATION
+            // seconds have passed.
+            if (front_k.txn_expiration_time + GARBAGE_COLLECTION_INTERVAL_SECS_AFTER_TXN_EXPIRATION < current_time) {
+                bucket.nonces_ordered_by_exp_time.pop_front();
+                bucket.nonce_to_exp_time_map.remove(&NonceKey {
+                    sender_address: front_k.sender_address,
+                    nonce: front_k.nonce,
+                });
+            } else {
+                break;
+            };
+            i = i + 1;
+        };
+
+        // Insert the (address, nonce) pair in the bucket.
+        let nonce_key_with_exp_time = NonceKeyWithExpTime {
+            txn_expiration_time,
+            sender_address,
+            nonce,
+        };
+        bucket.nonces_ordered_by_exp_time.add(nonce_key_with_exp_time, true);
+        bucket.nonce_to_exp_time_map.add(nonce_key, txn_expiration_time);
+        true
+    }
+
+    // Returns true if the input (address, nonce) pair doesn't exist in the nonce history.
+    // Returns false if the input (address, nonce) pair already exists in the nonce history.
+    public(friend) fun check_nonce(
+        sender_address: address,
+        nonce: u64,
+    ): bool acquires NonceHistory {
```

Comment:
hmm, doesn't this need to take in the txn_expiration , to have correct computation now?

since this wasn't caught by tests - let's add a few tests for corner cases, to verify check_nonce returns the same thing check_and_insert_nonce would

Reply:
This function isn't even used anywhere in the main logic. I just used in the tests to check whether an (addres, nonce) pair exists in the nonce history.

Reply:
I would then call it something like "check_if_nonce_exists_in_table" because it has different expiration rules than check_and_insert_nonce  (and make it private? )